xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authentication/xwiki-platform-security-authentication-default/src/main/java/org/xwiki/security/authentication/internal/resource/AuthenticationResourceReferenceHandler.java

```diff
@@ -0,0 +1,108 @@
+ */
+package org.xwiki.security.authentication.internal.resource;
+
+import java.util.Collections;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.context.Execution;
+import org.xwiki.context.ExecutionContext;
+import org.xwiki.resource.AbstractResourceReferenceHandler;
+import org.xwiki.resource.ResourceReference;
+import org.xwiki.resource.ResourceReferenceHandlerChain;
+import org.xwiki.resource.ResourceReferenceHandlerException;
+import org.xwiki.resource.ResourceType;
+import org.xwiki.security.authentication.api.AuthenticationResourceReference;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiContextInitializer;
+import com.xpn.xwiki.web.Utils;
+
+/**
+ * Default component to handle {@link AuthenticationResourceReference}.
+ *
+ * @version $Id$
+ * @since 13.1RC1
+ */
+@Component
+@Named(AuthenticationResourceReference.RESOURCE_TYPE_ID)
+@Singleton
+public class AuthenticationResourceReferenceHandler extends AbstractResourceReferenceHandler<ResourceType>
+{
+    @Inject
+    private XWikiContextInitializer xWikiContextInitializer;
+
+    @Inject
+    private Execution execution;
+
+    @Override
+    public List<ResourceType> getSupportedResourceReferences()
+    {
+        return Collections.singletonList(AuthenticationResourceReference.TYPE);
+    }
+
+    @Override
+    public void handle(ResourceReference reference, ResourceReferenceHandlerChain chain)
+        throws ResourceReferenceHandlerException
+    {
+        AuthenticationResourceReference authenticationResourceReference = (AuthenticationResourceReference) reference;
+
+        switch (authenticationResourceReference.getAction()) {
+            case FORGOT_USERNAME:
+                this.handleAction("forgotusername");
+                break;
+
+            case RESET_PASSWORD:
+                this.handleAction("resetpassword");
+                break;
+
+            default:
+                // nothing to do here.
+        }
+
+        // Be a good citizen, continue the chain, in case some lower-priority Handler has something to do for this
+        // Resource Reference.
+        chain.handleNext(reference);
+    }
+
+    private void handleAction(String templateName) throws ResourceReferenceHandlerException
+    {
+        ExecutionContext context = this.execution.getContext();
+        if (context == null) {
+            context = new ExecutionContext();
+        }
+        try {
+            XWikiContext xWikiContext = this.xWikiContextInitializer.initialize(context);
+            // We are directly relying on Utils#parseTemplate because we want the plugin manager to properly
+            // handle the javascript placeholders and it avoids duplicating code.
+            Utils.parseTemplate(templateName, true, xWikiContext);
```

Comment:
old api usage. Should probably use the TemplateManager.

Reply:
So I explained on the comment why I did not use the TemplateManager: as a matter of fact I tried first to use the template manager, but I needed more stuff to handle the JSX placeholders and I discovered that this methoud was already doing all the plumbing I needed.

Reply:
ok, I reviewed quickly and missed the comment. We'll need to do something at some point since we need to deprecate and remove `Utils.parseTemplate`. Either put that it does in TemplateManager somewhere, or introduce some new component.